be/src/connector/binlog_connector.h

```diff
@@ -87,13 +83,23 @@ class BinlogDataSource final : public StreamDataSource {
     StatusOr<TabletSharedPtr> _get_tablet();
     BinlogMetaFieldMap _build_binlog_meta_fields(ColumnId start_cid);
     StatusOr<Schema> _build_binlog_schema();
+    Status _prepare_non_stream_pipeline();
 
+    const PlanNodesConstants _column_name_constants;
     const BinlogDataSourceProvider* _provider;
     const TBinlogScanRange _scan_range;
     RuntimeState* _runtime_state = nullptr;
+    bool _is_stream_pipeline = false;
+
     TabletSharedPtr _tablet;
-    // TODO this will be used by BinlogReader
     Schema _binlog_read_schema;
+    BinlogReaderSharedPtr _binlog_reader;
+
+    // whether to need do a seek before read data
+    std::atomic<bool> _need_seek_binlog{true};
```

Comment:
Under what situation will there be concurrent access?

Reply:
`set_offset()` and `get_next()` may be in different threads?